### PR TITLE
Support backwards compatibility for v1beta1 custom metrics client

### LINKED ISSUE #68011: HPA no longer works with adapters exposing Custom Metrics v1beta1 API
Merging https://github.com/kubernetes/kubernetes/pull/64097 has broken e2e tests: https://k8s-testgrid.appspot.com/sig-autoscaling-hpa#gci-gce-autoscaling-hpa. Based on the logs it seems that HPA can no longer retrieve custom metrics provided by adapter (it still works for external metrics though).

Quickly looking through the code it looks like #64097 updated https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go#L115 to use Custom Metrics v1beta2 API. Since v1beta2 is not backward compatible with v1beta1 and there is no fallback to old API, I suspect HPA will no longer work with any pre-existing adapter implementing v1beta1 API (like the stackdriver adapter we use in tests). 

If my suspicion is correct I think this would be a huge regression - all existing adapters will stop working in 1.12. As per https://kubernetes.io/docs/reference/using-api/deprecation-policy/ we can't just drop support for v1beta1 - we need to support it for 9 months / 3 releases after officially deprecating it.

/kind bug
/priority critical-urgent
/sig autoscaling

cc: @kubernetes/sig-autoscaling-bugs @mwielgus @DirectXMan12 @damemi @kawych @piosz

### PR DESCRIPTION
**Pull request by damemi**
Fixes #68011. The custom metrics client was updated to expect v1beta2 which broke all existing adapters. This reverts that change and adds a conversion from v1beta1 to v1beta2 internally

**What this PR does / why we need it**:
Fixes a bug in #64097 which broke custom_metrics clients for adapters using v1beta1

**Which issue(s) this PR fixes**
Fixes #68011

**Special notes for your reviewer**:

**Release note**:
```
NONE
```

### CODE REVIEW

**Comment by DirectXMan12** [comment]
@kubernetes/sig-autoscaling-pr-reviews

File: staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go
Comment (MaciekPytel) [inline staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: I don't really understand REST client machinery well enough, so I may very well be wrong on this, but won't this cause the following call to fail for an adapter that actually implements v1beta2? 

We initially put the result into v1beta1.MetricValueList{} and only later apply the conversion. But I don't expect v1beta2 result to be able to deserialize into a v1beta1 value list as MetricValue type is not compatible between versions. Am I missing something here?

Comment (DirectXMan12) [reply staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: Adapters will have to implement both APIs for the time being, and then we'll have to do the switch over in a later release.

So adapters should start implementing v1beta2 now, then when we switch over next release (or the one after) here, they'll have support for it, and a given adapter will work with both versions of kube.  Then, the release after that, we'll be able to finally remove v1beta1.

(that's the only way I can see this working, without fallback logic, and I don't think there's a precedence for fallback logic since we stopped using unversioned clients).

Comment (MaciekPytel) [reply staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: I see your point. I wonder if we could address this by having a separate client for each version, discover which version is available and use the appropriate client in HPA?

I don't like the approach of switching the version at some point, mainly because I'm skeptical about every adapter implementing v1beta2 while it's not actually supported by k8s. I think we're effectively just pushing the problem to next release.

Comment (DirectXMan12) [reply staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: > I see your point. I wonder if we could address this by having a separate client for each version, discover which version is available and use the appropriate client in HPA?

I considered that, but it's way more complicated, and doesn't really seem to have a precendent in the rest of the codebase, IIRC (ever since we removed the unversioned clients).

> In the meantime every new adapter would have to implement a deprecated API.

Nobody needs to implement anything new except for the developers of kubernetes-incubator/custom-metrics-apiserver -- all adapters should get the APIs for free when they update.

Perhaps we should have @kubernetes/sig-api-machinery-pr-reviews weigh in?

Comment (deads2k) [reply staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: @lavalamp another concrete example of the utility of a hub an a hubbed client.

@DirectXMan12 if it were my code, I would generate two different clients.  Then given an apiserver, I would use discovery to choose from the available APIs to new up the "best" client.  Then, if necessary, I would provide adapter clients that could skin one version as another one.  It's pretty painful, but with only two versions, it isn't so bad.

Comment (DirectXMan12) [reply staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: @deads2k ack, that seems reasonable to me eventually.  do we want to try to do it less than a week before code freeze while the tests are broken, though?

@damemi your mission, should you choose to accept it ^

**Comment by damemi** [comment]
Cleaned up the fix and tested manually using a custom metrics adapter that expects v1beta1. Seems to work, just need to sort out the logic for determining the actual preferred version (instead of hardcoding v1beta1, which it is right now) but that's pretty much all there. Will update that and report back when this is ready for a final look before merge

**Comment by DirectXMan12** [comment]
SIG release: in case you're looking at this and want an update, and you can't reach me on Slack:

We either: a) need to merge the earlier version of this (relatively low risk)
 or b) need a bit more time (probably 1 or two days, depending on review latency) to finish and merge the fix suggested by @deads2k.  In either case, we need to merge *a* fix, as without it backwards compat with older CM API adapters is broken currently.

**Comment by damemi** [comment]
@DirectXMan12 sorry forgot to comment but I did finish pushing the changes David recommended the other day, and the tests look green right now (also tested locally). 

I just don't like the approach of having to manually loop through all of the APIGroups and versions to find one that matches, but the Discovery client doesn't really provide a much better way unless I'm missing it. We could add a new helper to Discovery that maybe only returns the requested APIGroup but that's really just shifting the work around. If it's not worth it, then this fix is ready from my end

**Comment by DirectXMan12** [comment]
I'll need to do a quick review.  I'll try to get to it today, but chances are we'll need a short exception

**Comment by liggitt** [comment]
> chances are we'll need a short exception

no exception needed for a bug fix restoring backwards compatibility prior to release, though a heads up to the release team would be appreciated, I'm sure.

Tagging the issue and/or into the milestone should get it on their radar as well.

File: staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go
Comment (DirectXMan12) [inline staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: this should be taking a `AvailableMetricsAPIFunc`, not an `APIVersionsFromDiscovery`.

Comment (DirectXMan12) [reply staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: ditto below

File: staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go
Comment (DirectXMan12) [inline staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: this actual implementation should be filtering the groups by the metrics API -- it shouldn't be down the code in the metrics client.  This allows us to test the two independently.

Comment (DirectXMan12) [reply staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go]: (i.e. I'd lift the "find the metrics API group" out of the converter).

File: cmd/kube-controller-manager/app/autoscaling.go
Comment (DirectXMan12) [inline cmd/kube-controller-manager/app/autoscaling.go]: this is going to be checking discovery for *every* metrics request.  We should be caching the information and periodically rechecking it, or rechecking if we don't have any results.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
needs a few tweaks -- comments inline

**Comment by DirectXMan12** [comment]
/kind bug
/priority critical-urgent

(this is a non-trivial regression in compatibility)

**Comment by DirectXMan12** [comment]
/remove-sig architecture
/remove-sig networking
/remove-sig 
/remove-area kubelet 
/remove-sig node
/remove-sig testing

the autolabeller seems to have guessed wrong...

**Comment by deads2k** [comment]
controller changes look ok.  You'll need an api-approver to put the API back.

/approve

**Comment by smarterclayton** [comment]
/approve

API changes to optional

**Comment by nikhita** [comment]
@luxas looks like this should contain a release note?

**Comment by DirectXMan12** [comment]
@nikhita no, it shouldn't since it's a fix for an unintentional issue introduced in a previous PR this release.  This now works as noted in the original PR's release note.

**Comment by nikhita** [comment]
@DirectXMan12 gotcha, thanks for the update!

**Pull request merged**
